{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "b5df5291",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sqlite3\n",
    "import pandas as pd\n",
    "import requests\n",
    "import time\n",
    "from bs4 import BeautifulSoup\n",
    "\n",
    "# sqlitebrowser\n",
    "\n",
    "# create diagram https://dbdiagram.io/d"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6cd0eb14",
   "metadata": {},
   "source": [
    "# Create sql from csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7f90498d",
   "metadata": {},
   "outputs": [],
   "source": [
    "genstudio = pd.read_csv('./data/genstudio.csv', dtype={'Position':str})\n",
    "metadata = pd.read_csv('./data/metadata.csv')\n",
    "\n",
    "#print(genstudio.shape, metadata.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "9bbaeb6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_scheme_df(df):\n",
    "    '''\n",
    "    Create SQL scheme from DataFrame colnames.\n",
    "    Return SQL scheme and list with processed colnames.\n",
    "    '''\n",
    "    col_names, types = df.columns, df.dtypes\n",
    "    scheme_df = ['row_id INTEGER PRIMARY KEY']\n",
    "    new_names = ['row_id']\n",
    "\n",
    "\n",
    "    for name, contain in zip(col_names[1:], types[1:]):\n",
    "        name = name.replace(\"-\", \"\").replace(\"  \", \" \").replace(\" \", \"_\")\n",
    "        new_names.append(name)\n",
    "\n",
    "        if 'int' in str(contain):\n",
    "            scheme_df.append(' '.join([name, 'INTEGER']))\n",
    "        \n",
    "        elif 'float' in str(contain):\n",
    "            scheme_df.append(' '.join([name, 'REAL']))\n",
    "        \n",
    "        else:\n",
    "            scheme_df.append(' '.join([name, 'TEXT']))\n",
    "            \n",
    "    scheme_df = ', '.join(scheme_df)\n",
    "\n",
    "    return scheme_df, new_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "fc8deac9",
   "metadata": {},
   "outputs": [],
   "source": [
    "scheme_genstudio, new_names = get_scheme_df(genstudio)\n",
    "genstudio.columns = new_names\n",
    "\n",
    "scheme_metadata, new_names = get_scheme_df(metadata)\n",
    "metadata.columns = new_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "df71bb08",
   "metadata": {},
   "outputs": [],
   "source": [
    "def bid_DB(genstudio, metadata, scheme_genstudio, scheme_metadata, path='./data/bid_DB.db'):\n",
    "    \n",
    "    connection = sqlite3.connect(path)\n",
    "\n",
    "    connection.execute('''CREATE TABLE IF NOT EXISTS SNP_data (\n",
    "    Sample_ID TEXT, \n",
    "    SNP_Name TEXT, \n",
    "    SNP TEXT, \n",
    "    Position TEXT, \n",
    "    Chr TEXT,\n",
    "    sex TEXT)''')\n",
    "\n",
    "    connection.execute(f'''CREATE TABLE IF NOT EXISTS metadata ({scheme_metadata},\n",
    "                        FOREIGN KEY (sex) REFERENCES SNP_data (sex))''')\n",
    "\n",
    "    connection.execute(f'''CREATE TABLE IF NOT EXISTS genstudio ({scheme_genstudio},\n",
    "    FOREIGN KEY (Sample_ID) REFERENCES metadata (dna_chip_id),\n",
    "    FOREIGN KEY (Sample_ID) REFERENCES metadata (Sample_ID),\n",
    "    FOREIGN KEY (SNP_Name) REFERENCES metadata (SNP_Name),\n",
    "    FOREIGN KEY (SNP) REFERENCES metadata (SNP),\n",
    "    FOREIGN KEY (Position) REFERENCES metadata (Position),\n",
    "    FOREIGN KEY (Chr) REFERENCES metadata (Chr))''')\n",
    "\n",
    "    connection.commit()\n",
    "\n",
    "    genstudio.to_sql('genstudio', connection, if_exists='append', index=False)\n",
    "    metadata.to_sql('metadata', connection, if_exists='append', index=False)\n",
    "\n",
    "    connection.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "221c57cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "path='./data/bid_DB.db'\n",
    "bid_DB(genstudio, metadata, scheme_genstudio, scheme_metadata, path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "382107d6",
   "metadata": {},
   "source": [
    "# Selection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "706d6e43",
   "metadata": {},
   "outputs": [],
   "source": [
    "connection = sqlite3.connect('./data/bid_DB.db')\n",
    "\n",
    "query = '''SELECT metadata.dna_chip_id, genstudio.SNP_Name, genstudio.SNP, genstudio.Position, genstudio.Chr, \n",
    "           metadata.sex\n",
    "           FROM genstudio, metadata\n",
    "           WHERE metadata.dna_chip_id = genstudio.Sample_ID AND genstudio.SNP = '[A/G]'\n",
    "'''\n",
    "\n",
    "res = connection.execute(query)\n",
    "\n",
    "print(len(res.fetchall()))\n",
    "\n",
    "connection.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38b6fe03",
   "metadata": {},
   "source": [
    "## Add selection to DB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "3931281d",
   "metadata": {},
   "outputs": [],
   "source": [
    "connection = sqlite3.connect('./data/bid_DB.db')\n",
    "\n",
    "query = '''INSERT INTO SNP_data (Sample_ID, SNP_Name, SNP, Position, Chr, sex)\n",
    "SELECT genstudio.Sample_ID, genstudio.SNP_Name, genstudio.SNP, genstudio.Position, genstudio.Chr, \n",
    "metadata.sex \n",
    "FROM genstudio, metadata \n",
    "WHERE metadata.dna_chip_id = genstudio.Sample_ID'''\n",
    "\n",
    "connection.execute(query)\n",
    "connection.commit()\n",
    "connection.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b57212e",
   "metadata": {},
   "source": [
    "# Create and add from web"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "id": "b0d791cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_genes():\n",
    "    \n",
    "    GENE_URL = 'https://genomics.senescence.info/genes/allgenes.php'\n",
    "    \n",
    "    resp = requests.get(GENE_URL)\n",
    "    soup = BeautifulSoup(resp.content, 'lxml')\n",
    "    \n",
    "    genes_list = []\n",
    "    \n",
    "    table = soup.find(\"table\", {\"class\":\"results-table\"} )\n",
    "    rows = table.find_all('tr')\n",
    "\n",
    "    for row in rows:\n",
    "        cols = row.find_all('td')\n",
    "        cols = [ele.text.strip() for ele in cols]\n",
    "        \n",
    "        if cols:    \n",
    "            genes_list.append(cols[0])\n",
    "    \n",
    "    return genes_list\n",
    "\n",
    "\n",
    "def snp_search(gene_name):\n",
    "    '''\n",
    "    Search first 20 SNP in ncbi snp database. Return soup object.\n",
    "    '''\n",
    "    SNP_URL = 'https://www.ncbi.nlm.nih.gov/snp/'\n",
    "    query_str = {'term': gene_name + '[Gene Name]'}\n",
    "    \n",
    "    resp = requests.get(SNP_URL, params=query_str)\n",
    "    soup = BeautifulSoup(resp.content, 'lxml')\n",
    "    \n",
    "    return soup\n",
    "\n",
    "\n",
    "def parse_soup(soup):\n",
    "    '''\n",
    "    Extract SNP ID from soup.\n",
    "    Return list with rs-id.\n",
    "    '''\n",
    "    rs_id = []\n",
    "    \n",
    "    for result in soup.find_all('div', {'class': 'rprt'}):\n",
    "        rs_id.append(result.find('div', {'class': 'supp'}).find('a').text)\n",
    "        \n",
    "    return rs_id\n",
    "\n",
    "\n",
    "def get_id_list(genes_list):\n",
    "    '''\n",
    "    Search first 20 SNP in ncbi snp database and extract SNP ID from soup.\n",
    "    Wait 5 seconds between get requests.\n",
    "    Return lists with SNP ID`s for all genes.\n",
    "    '''\n",
    "    id_list = []\n",
    "    \n",
    "    for gene_name in genes_list:\n",
    "        #time.sleep(5)\n",
    "        \n",
    "        soup = snp_search(str(gene_name))\n",
    "        rs_id = parse_soup(soup)\n",
    "\n",
    "        if rs_id:\n",
    "            id_list.append(rs_id)\n",
    "            \n",
    "    return id_list\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "id": "c8ae2435",
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_to_sql(snp_name, position, chromosome, alleles, db_build, assembly_build,\n",
    "               clin_sig, n_publications):\n",
    "    '''\n",
    "    Open connection to SNP_human DB, add some info to it and close connection.\n",
    "    Return nothing.\n",
    "    '''\n",
    "    connection = sqlite3.connect('./data/SNP_human.db')\n",
    "    \n",
    "    query_1 = '''INSERT INTO SNP_data (SNP_Name, Alleles, Position, Chr, dbSNP_Build, Assembly_Build)\n",
    "                                       VALUES (?, ?, ?, ?, ?, ?)'''\n",
    "    \n",
    "    connection.execute(query_1, [snp_name, alleles, position, chromosome, db_build, assembly_build])\n",
    "    \n",
    "    query_2 = '''INSERT INTO clin_SNP (SNP_Name, Clinical_Significance, N_Publications)\n",
    "                                       VALUES (?, ?, ?)'''\n",
    "    \n",
    "    connection.execute(query_2, [snp_name, clin_sig, n_publications]) \n",
    "    \n",
    "    connection.commit()\n",
    "    connection.close()\n",
    "\n",
    "\n",
    "def get_snp_info(red_id_list):\n",
    "    '''\n",
    "    Get info about SNP from soup object and add it to the SQL DB.\n",
    "    Wait 5 seconds between get requests.\n",
    "    Return 'DONE' when done.\n",
    "    '''\n",
    "    i = 1\n",
    "    \n",
    "    for snp_name in red_id_list:\n",
    "        \n",
    "        print(i)\n",
    "        \n",
    "        #time.sleep(5)\n",
    "        \n",
    "        req_url = f'https://www.ncbi.nlm.nih.gov/snp/{snp_name}#clinical_significance'\n",
    "\n",
    "        resp = requests.get(req_url, params={'horizontal_tab': 'true'})\n",
    "        soup = BeautifulSoup(resp.content, 'lxml')\n",
    "\n",
    "        db_build = soup.find('div', {'class': 'accession usa-width-one-third'}).text.split()[2]\n",
    "\n",
    "        snp_info = soup.find_all('dl', {'class': 'usa-width-one-half'})\n",
    "        dd_info = snp_info[0].find_all('dd')\n",
    "        build_chr_pos = dd_info[1].find_all('span') \n",
    "\n",
    "        assembly_build = build_chr_pos[1].text.strip()\n",
    "        position = build_chr_pos[0].text.split(':')[1]\n",
    "        chromosome = build_chr_pos[0].text.split(':')[0]\n",
    "\n",
    "        alleles = ''.join(dd_info[2].text.split())\n",
    "    \n",
    "        clin_sig = snp_info[1].find('dd').text.strip()\n",
    "        \n",
    "        publ_info = snp_info[1].find('a', {'id': 'snp_pub_count'})\n",
    "        \n",
    "        if publ_info:\n",
    "            n_publications = publ_info.text.split()[0]\n",
    "        else:\n",
    "            n_publications=0\n",
    "        \n",
    "        add_to_sql(snp_name,\n",
    "                   position,\n",
    "                   chromosome,\n",
    "                   alleles,\n",
    "                   db_build,\n",
    "                   assembly_build,\n",
    "                   clin_sig,\n",
    "                   n_publications)\n",
    "        i += 1\n",
    "    \n",
    "    return 'DONE'\n",
    "\n",
    "\n",
    "def create_SNP_db():\n",
    "    '''\n",
    "    Create DB with two tables with info about human genes SNP.\n",
    "    Gets information about the name of human genes from the \"genomics.senescence.info\" site.\n",
    "    When searching for SNPs for each gene in SNP NCBI, returns no more than 20 SNPs.\n",
    "    Write found SNPs to the SNP_ids.txt file.\n",
    "    Search info about each SNP in SNP NCBI.\n",
    "    Add info about found SNPs to SNP_human.db.\n",
    "    Return DONE when done.\n",
    "    \n",
    "    '''\n",
    "    \n",
    "    connection = sqlite3.connect('./data/SNP_human.db')\n",
    "\n",
    "    connection.execute('''CREATE TABLE IF NOT EXISTS SNP_data ( \n",
    "                           SNP_Name TEXT, \n",
    "                           Alleles TEXT, \n",
    "                           Position TEXT, \n",
    "                           Chr TEXT,\n",
    "                           dbSNP_Build TEXT,\n",
    "                           Assembly_Build TEXT)''')\n",
    "\n",
    "    connection.execute('''CREATE TABLE IF NOT EXISTS clin_SNP (\n",
    "                           SNP_Name TEXT,\n",
    "                           Clinical_Significance TEXT,\n",
    "                           N_Publications TEXT,\n",
    "                           FOREIGN KEY (SNP_Name) REFERENCES SNP_data (SNP_Name))''')\n",
    "\n",
    "    connection.commit()\n",
    "    connection.close()\n",
    "    ''' # Done if file\n",
    "    genes_list = get_genes()\n",
    "\n",
    "    id_list = get_id_list(genes_list)\n",
    "    red_id_list = list(set([j for sub in id_list for j in sub]))\n",
    "\n",
    "    with open('./data/SNP_ids.txt', 'w') as output_file:\n",
    "        for rsid in red_id_list:\n",
    "            output_file.write(rsid + '\\n')\n",
    "    '''\n",
    "    \n",
    "    with open('./data/SNP_ids.txt') as f:\n",
    "        red_id_list = f.read().splitlines()\n",
    "    \n",
    "    return get_snp_info(red_id_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6ff6c71",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n",
      "2\n",
      "3\n",
      "4\n",
      "5\n",
      "6\n",
      "7\n",
      "8\n",
      "9\n",
      "10\n",
      "11\n",
      "12\n",
      "13\n",
      "14\n",
      "15\n",
      "16\n",
      "17\n",
      "18\n",
      "19\n",
      "20\n",
      "21\n",
      "22\n",
      "23\n",
      "24\n",
      "25\n",
      "26\n",
      "27\n",
      "28\n",
      "29\n",
      "30\n",
      "31\n",
      "32\n",
      "33\n",
      "34\n",
      "35\n",
      "36\n",
      "37\n",
      "38\n",
      "39\n",
      "40\n",
      "41\n",
      "42\n",
      "43\n",
      "44\n",
      "45\n",
      "46\n",
      "47\n",
      "48\n",
      "49\n",
      "50\n",
      "51\n",
      "52\n",
      "53\n",
      "54\n",
      "55\n",
      "56\n",
      "57\n",
      "58\n",
      "59\n",
      "60\n",
      "61\n",
      "62\n",
      "63\n",
      "64\n",
      "65\n",
      "66\n",
      "67\n",
      "68\n",
      "69\n",
      "70\n",
      "71\n",
      "72\n",
      "73\n",
      "74\n",
      "75\n",
      "76\n",
      "77\n",
      "78\n",
      "79\n",
      "80\n",
      "81\n",
      "82\n",
      "83\n",
      "84\n",
      "85\n",
      "86\n",
      "87\n",
      "88\n",
      "89\n",
      "90\n",
      "91\n",
      "92\n",
      "93\n",
      "94\n",
      "95\n",
      "96\n",
      "97\n",
      "98\n",
      "99\n",
      "100\n",
      "101\n",
      "102\n",
      "103\n",
      "104\n",
      "105\n",
      "106\n",
      "107\n",
      "108\n",
      "109\n",
      "110\n",
      "111\n",
      "112\n",
      "113\n",
      "114\n",
      "115\n",
      "116\n",
      "117\n",
      "118\n",
      "119\n",
      "120\n",
      "121\n",
      "122\n",
      "123\n",
      "124\n",
      "125\n",
      "126\n",
      "127\n",
      "128\n",
      "129\n",
      "130\n",
      "131\n",
      "132\n",
      "133\n",
      "134\n",
      "135\n",
      "136\n",
      "137\n",
      "138\n",
      "139\n",
      "140\n",
      "141\n",
      "142\n",
      "143\n",
      "144\n",
      "145\n",
      "146\n",
      "147\n",
      "148\n",
      "149\n",
      "150\n",
      "151\n",
      "152\n",
      "153\n",
      "154\n",
      "155\n",
      "156\n",
      "157\n",
      "158\n",
      "159\n",
      "160\n",
      "161\n",
      "162\n",
      "163\n",
      "164\n",
      "165\n",
      "166\n",
      "167\n",
      "168\n",
      "169\n",
      "170\n",
      "171\n",
      "172\n",
      "173\n",
      "174\n",
      "175\n",
      "176\n",
      "177\n",
      "178\n",
      "179\n",
      "180\n",
      "181\n",
      "182\n",
      "183\n",
      "184\n",
      "185\n",
      "186\n",
      "187\n",
      "188\n",
      "189\n",
      "190\n",
      "191\n",
      "192\n",
      "193\n",
      "194\n",
      "195\n",
      "196\n",
      "197\n",
      "198\n",
      "199\n",
      "200\n",
      "201\n",
      "202\n",
      "203\n",
      "204\n",
      "205\n",
      "206\n",
      "207\n",
      "208\n",
      "209\n",
      "210\n",
      "211\n",
      "212\n",
      "213\n",
      "214\n",
      "215\n",
      "216\n",
      "217\n",
      "218\n",
      "219\n",
      "220\n",
      "221\n",
      "222\n",
      "223\n",
      "224\n",
      "225\n",
      "226\n",
      "227\n",
      "228\n",
      "229\n",
      "230\n",
      "231\n",
      "232\n",
      "233\n",
      "234\n",
      "235\n",
      "236\n",
      "237\n",
      "238\n",
      "239\n",
      "240\n",
      "241\n",
      "242\n",
      "243\n",
      "244\n",
      "245\n",
      "246\n",
      "247\n",
      "248\n",
      "249\n",
      "250\n",
      "251\n",
      "252\n",
      "253\n",
      "254\n",
      "255\n",
      "256\n",
      "257\n",
      "258\n",
      "259\n",
      "260\n",
      "261\n",
      "262\n",
      "263\n",
      "264\n",
      "265\n",
      "266\n",
      "267\n",
      "268\n",
      "269\n",
      "270\n",
      "271\n",
      "272\n",
      "273\n",
      "274\n",
      "275\n",
      "276\n",
      "277\n",
      "278\n",
      "279\n",
      "280\n",
      "281\n",
      "282\n",
      "283\n",
      "284\n",
      "285\n",
      "286\n",
      "287\n",
      "288\n",
      "289\n",
      "290\n",
      "291\n",
      "292\n",
      "293\n",
      "294\n",
      "295\n",
      "296\n",
      "297\n",
      "298\n",
      "299\n",
      "300\n",
      "301\n",
      "302\n",
      "303\n",
      "304\n",
      "305\n",
      "306\n",
      "307\n",
      "308\n",
      "309\n",
      "310\n",
      "311\n",
      "312\n",
      "313\n",
      "314\n",
      "315\n",
      "316\n",
      "317\n",
      "318\n",
      "319\n",
      "320\n",
      "321\n",
      "322\n",
      "323\n",
      "324\n",
      "325\n",
      "326\n",
      "327\n",
      "328\n",
      "329\n",
      "330\n",
      "331\n",
      "332\n",
      "333\n",
      "334\n",
      "335\n",
      "336\n",
      "337\n",
      "338\n",
      "339\n",
      "340\n",
      "341\n",
      "342\n",
      "343\n",
      "344\n",
      "345\n",
      "346\n",
      "347\n",
      "348\n",
      "349\n",
      "350\n",
      "351\n",
      "352\n",
      "353\n",
      "354\n",
      "355\n",
      "356\n",
      "357\n",
      "358\n",
      "359\n",
      "360\n",
      "361\n",
      "362\n",
      "363\n",
      "364\n",
      "365\n",
      "366\n",
      "367\n",
      "368\n",
      "369\n",
      "370\n",
      "371\n",
      "372\n",
      "373\n",
      "374\n",
      "375\n",
      "376\n",
      "377\n",
      "378\n",
      "379\n",
      "380\n",
      "381\n",
      "382\n",
      "383\n",
      "384\n",
      "385\n",
      "386\n",
      "387\n",
      "388\n",
      "389\n",
      "390\n",
      "391\n",
      "392\n",
      "393\n",
      "394\n",
      "395\n",
      "396\n",
      "397\n",
      "398\n",
      "399\n",
      "400\n",
      "401\n",
      "402\n",
      "403\n",
      "404\n",
      "405\n",
      "406\n",
      "407\n",
      "408\n",
      "409\n",
      "410\n",
      "411\n",
      "412\n",
      "413\n",
      "414\n",
      "415\n",
      "416\n",
      "417\n",
      "418\n",
      "419\n",
      "420\n",
      "421\n",
      "422\n",
      "423\n",
      "424\n",
      "425\n",
      "426\n",
      "427\n",
      "428\n",
      "429\n",
      "430\n",
      "431\n",
      "432\n",
      "433\n",
      "434\n",
      "435\n",
      "436\n",
      "437\n",
      "438\n",
      "439\n",
      "440\n",
      "441\n",
      "442\n",
      "443\n",
      "444\n",
      "445\n",
      "446\n",
      "447\n",
      "448\n",
      "449\n",
      "450\n",
      "451\n",
      "452\n",
      "453\n",
      "454\n",
      "455\n",
      "456\n",
      "457\n",
      "458\n",
      "459\n",
      "460\n",
      "461\n",
      "462\n",
      "463\n",
      "464\n",
      "465\n",
      "466\n",
      "467\n",
      "468\n",
      "469\n",
      "470\n",
      "471\n",
      "472\n",
      "473\n",
      "474\n",
      "475\n",
      "476\n",
      "477\n",
      "478\n",
      "479\n",
      "480\n",
      "481\n",
      "482\n",
      "483\n",
      "484\n",
      "485\n",
      "486\n",
      "487\n",
      "488\n",
      "489\n",
      "490\n",
      "491\n",
      "492\n",
      "493\n",
      "494\n",
      "495\n",
      "496\n",
      "497\n",
      "498\n",
      "499\n",
      "500\n",
      "501\n",
      "502\n",
      "503\n",
      "504\n",
      "505\n",
      "506\n",
      "507\n",
      "508\n",
      "509\n",
      "510\n",
      "511\n",
      "512\n",
      "513\n",
      "514\n",
      "515\n",
      "516\n",
      "517\n",
      "518\n",
      "519\n",
      "520\n",
      "521\n",
      "522\n",
      "523\n",
      "524\n",
      "525\n",
      "526\n",
      "527\n",
      "528\n",
      "529\n",
      "530\n",
      "531\n",
      "532\n",
      "533\n",
      "534\n",
      "535\n",
      "536\n",
      "537\n",
      "538\n",
      "539\n",
      "540\n"
     ]
    }
   ],
   "source": [
    "create_SNP_db()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c442a2c3",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
