{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b5df5291",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sqlite3\n",
    "import pandas as pd\n",
    "import requests\n",
    "import time\n",
    "from bs4 import BeautifulSoup\n",
    "\n",
    "# sqlitebrowser\n",
    "# create diagram https://dbdiagram.io/d"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6cd0eb14",
   "metadata": {},
   "source": [
    "# Create sql from csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7f90498d",
   "metadata": {},
   "outputs": [],
   "source": [
    "genstudio = pd.read_csv('./data/genstudio.csv', dtype={'Position':str})\n",
    "metadata = pd.read_csv('./data/metadata.csv')\n",
    "\n",
    "#print(genstudio.shape, metadata.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9bbaeb6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_scheme_df(df):\n",
    "    '''\n",
    "    Create SQL scheme from DataFrame colnames.\n",
    "    Return SQL scheme and list with processed colnames.\n",
    "    '''\n",
    "    col_names, types = df.columns, df.dtypes\n",
    "    scheme_df = ['row_id INTEGER PRIMARY KEY']\n",
    "    new_names = ['row_id']\n",
    "\n",
    "\n",
    "    for name, contain in zip(col_names[1:], types[1:]):\n",
    "        name = name.replace(\"-\", \"\").replace(\"  \", \" \").replace(\" \", \"_\")\n",
    "        new_names.append(name)\n",
    "\n",
    "        if 'int' in str(contain):\n",
    "            scheme_df.append(' '.join([name, 'INTEGER']))\n",
    "        \n",
    "        elif 'float' in str(contain):\n",
    "            scheme_df.append(' '.join([name, 'REAL']))\n",
    "        \n",
    "        else:\n",
    "            scheme_df.append(' '.join([name, 'TEXT']))\n",
    "            \n",
    "    scheme_df = ', '.join(scheme_df)\n",
    "\n",
    "    return scheme_df, new_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "fc8deac9",
   "metadata": {},
   "outputs": [],
   "source": [
    "scheme_genstudio, new_names = get_scheme_df(genstudio)\n",
    "genstudio.columns = new_names\n",
    "\n",
    "scheme_metadata, new_names = get_scheme_df(metadata)\n",
    "metadata.columns = new_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "df71bb08",
   "metadata": {},
   "outputs": [],
   "source": [
    "def bid_DB(genstudio, metadata, scheme_genstudio, scheme_metadata, path='./data/bid_DB.db'):\n",
    "    \n",
    "    connection = sqlite3.connect(path)\n",
    "\n",
    "    connection.execute('''CREATE TABLE IF NOT EXISTS SNP_data (\n",
    "    Sample_ID TEXT, \n",
    "    SNP_Name TEXT, \n",
    "    SNP TEXT, \n",
    "    Position TEXT, \n",
    "    Chr TEXT,\n",
    "    sex TEXT)''')\n",
    "\n",
    "    connection.execute(f'''CREATE TABLE IF NOT EXISTS metadata ({scheme_metadata},\n",
    "                        FOREIGN KEY (sex) REFERENCES SNP_data (sex))''')\n",
    "\n",
    "    connection.execute(f'''CREATE TABLE IF NOT EXISTS genstudio ({scheme_genstudio},\n",
    "    FOREIGN KEY (Sample_ID) REFERENCES metadata (dna_chip_id),\n",
    "    FOREIGN KEY (Sample_ID) REFERENCES metadata (Sample_ID),\n",
    "    FOREIGN KEY (SNP_Name) REFERENCES metadata (SNP_Name),\n",
    "    FOREIGN KEY (SNP) REFERENCES metadata (SNP),\n",
    "    FOREIGN KEY (Position) REFERENCES metadata (Position),\n",
    "    FOREIGN KEY (Chr) REFERENCES metadata (Chr))''')\n",
    "\n",
    "    connection.commit()\n",
    "\n",
    "    genstudio.to_sql('genstudio', connection, if_exists='append', index=False)\n",
    "    metadata.to_sql('metadata', connection, if_exists='append', index=False)\n",
    "\n",
    "    connection.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "221c57cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "path='./data/bid_DB.db'\n",
    "bid_DB(genstudio, metadata, scheme_genstudio, scheme_metadata, path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ca1423f",
   "metadata": {},
   "source": [
    "# Selection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9266803c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "855870\n"
     ]
    }
   ],
   "source": [
    "connection = sqlite3.connect('./data/bid_DB.db')\n",
    "\n",
    "query = '''SELECT Sample_ID, SNP, Chr\n",
    "           FROM genstudio\n",
    "           WHERE SNP = '[A/G]'\n",
    "'''\n",
    "\n",
    "res = connection.execute(query)\n",
    "\n",
    "print(len(res.fetchall()))\n",
    "\n",
    "connection.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "054695ee",
   "metadata": {},
   "source": [
    "## Add selection to DB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "3931281d",
   "metadata": {},
   "outputs": [],
   "source": [
    "connection = sqlite3.connect('./data/bid_DB.db')\n",
    "\n",
    "query = '''INSERT INTO SNP_data (Sample_ID, SNP_Name, SNP, Position, Chr, sex)\n",
    "SELECT genstudio.Sample_ID, genstudio.SNP_Name, genstudio.SNP, genstudio.Position, genstudio.Chr, \n",
    "metadata.sex \n",
    "FROM genstudio, metadata \n",
    "WHERE metadata.dna_chip_id = genstudio.Sample_ID'''\n",
    "\n",
    "connection.execute(query)\n",
    "connection.commit()\n",
    "connection.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4056f948",
   "metadata": {},
   "source": [
    "# Create and add from web"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4abb85e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_genes():\n",
    "    \n",
    "    GENE_URL = 'https://genomics.senescence.info/genes/allgenes.php'\n",
    "    \n",
    "    resp = requests.get(GENE_URL)\n",
    "    soup = BeautifulSoup(resp.content, 'lxml')\n",
    "    \n",
    "    genes_list = []\n",
    "    \n",
    "    table = soup.find(\"table\", {\"class\":\"results-table\"} )\n",
    "    rows = table.find_all('tr')\n",
    "\n",
    "    for row in rows:\n",
    "        cols = row.find_all('td')\n",
    "        cols = [ele.text.strip() for ele in cols]\n",
    "        \n",
    "        if cols:    \n",
    "            genes_list.append(cols[0])\n",
    "    \n",
    "    return genes_list\n",
    "\n",
    "\n",
    "def snp_search(gene_name):\n",
    "    '''\n",
    "    Search first 20 SNP in ncbi snp database. Return soup object.\n",
    "    '''\n",
    "    SNP_URL = 'https://www.ncbi.nlm.nih.gov/snp/'\n",
    "    query_str = {'term': gene_name + '[Gene Name]'}\n",
    "    \n",
    "    resp = requests.get(SNP_URL, params=query_str)\n",
    "    soup = BeautifulSoup(resp.content, 'lxml')\n",
    "    \n",
    "    return soup\n",
    "\n",
    "\n",
    "def parse_soup(soup):\n",
    "    '''\n",
    "    Extract SNP ID from soup.\n",
    "    Return list with rs-id.\n",
    "    '''\n",
    "    rs_id = []\n",
    "    \n",
    "    for result in soup.find_all('div', {'class': 'rprt'}):\n",
    "        rs_id.append(result.find('div', {'class': 'supp'}).find('a').text)\n",
    "        \n",
    "    return rs_id\n",
    "\n",
    "\n",
    "def get_id_list(genes_list):\n",
    "    '''\n",
    "    Search first 20 SNP in ncbi snp database and extract SNP ID from soup.\n",
    "    Wait 5 seconds between get requests.\n",
    "    Return lists with SNP ID`s for all genes.\n",
    "    '''\n",
    "    id_list = []\n",
    "    \n",
    "    for gene_name in genes_list:\n",
    "        time.sleep(1)\n",
    "        \n",
    "        soup = snp_search(str(gene_name))\n",
    "        rs_id = parse_soup(soup)\n",
    "\n",
    "        if rs_id:\n",
    "            id_list.append(rs_id)\n",
    "            \n",
    "    return id_list\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73005e32",
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_to_sql(snp_name, position, chromosome, alleles, db_build, assembly_build,\n",
    "               clin_sig, n_publications):\n",
    "    '''\n",
    "    Open connection to SNP_human DB, add some info to it and close connection.\n",
    "    Return nothing.\n",
    "    '''\n",
    "    connection = sqlite3.connect('./data/SNP_human.db')\n",
    "    \n",
    "    query_1 = '''INSERT INTO SNP_data (SNP_Name, Alleles, Position, Chr, dbSNP_Build, Assembly_Build)\n",
    "                                       VALUES (?, ?, ?, ?, ?, ?)'''\n",
    "    \n",
    "    connection.execute(query_1, [snp_name, alleles, position, chromosome, db_build, assembly_build])\n",
    "    \n",
    "    query_2 = '''INSERT INTO clin_SNP (SNP_Name, Clinical_Significance, N_Publications)\n",
    "                                       VALUES (?, ?, ?)'''\n",
    "    \n",
    "    connection.execute(query_2, [snp_name, clin_sig, n_publications]) \n",
    "    \n",
    "    connection.commit()\n",
    "    connection.close()\n",
    "\n",
    "\n",
    "def get_snp_info(red_id_list):\n",
    "    '''\n",
    "    Get info about SNP from soup object and add it to the SQL DB.\n",
    "    Wait 5 seconds between get requests.\n",
    "    Return 'DONE' when done.\n",
    "    '''\n",
    "    i = 1\n",
    "    \n",
    "    for snp_name in red_id_list:\n",
    "        \n",
    "        time.sleep(1)\n",
    "        \n",
    "        req_url = f'https://www.ncbi.nlm.nih.gov/snp/{snp_name}#clinical_significance'\n",
    "\n",
    "        resp = requests.get(req_url, params={'horizontal_tab': 'true'})\n",
    "        soup = BeautifulSoup(resp.content, 'lxml')\n",
    "\n",
    "        db_build = soup.find('div', {'class': 'accession usa-width-one-third'}).text.split()[2]\n",
    "\n",
    "        snp_info = soup.find_all('dl', {'class': 'usa-width-one-half'})\n",
    "        dd_info = snp_info[0].find_all('dd')\n",
    "        build_chr_pos = dd_info[1].find_all('span') \n",
    "\n",
    "        assembly_build = build_chr_pos[1].text.strip()\n",
    "        position = build_chr_pos[0].text.split(':')[1]\n",
    "        chromosome = build_chr_pos[0].text.split(':')[0]\n",
    "\n",
    "        alleles = ''.join(dd_info[2].text.split())\n",
    "    \n",
    "        clin_sig = snp_info[1].find('dd').text.strip()\n",
    "        \n",
    "        publ_info = snp_info[1].find('a', {'id': 'snp_pub_count'})\n",
    "        \n",
    "        if publ_info:\n",
    "            n_publications = publ_info.text.split()[0]\n",
    "        else:\n",
    "            n_publications=0\n",
    "        \n",
    "        add_to_sql(snp_name,\n",
    "                   position,\n",
    "                   chromosome,\n",
    "                   alleles,\n",
    "                   db_build,\n",
    "                   assembly_build,\n",
    "                   clin_sig,\n",
    "                   n_publications)\n",
    "\n",
    "    return 'DONE'\n",
    "\n",
    "\n",
    "def create_SNP_db():\n",
    "    '''\n",
    "    Create DB with two tables with info about human genes SNP.\n",
    "    Gets information about the name of human genes from the \"genomics.senescence.info\" site.\n",
    "    When searching for SNPs for each gene in SNP NCBI, returns no more than 20 SNPs.\n",
    "    Write found SNPs to the SNP_ids.txt file.\n",
    "    Search info about each SNP in SNP NCBI.\n",
    "    Add info about found SNPs to SNP_human.db.\n",
    "    Return DONE when done.\n",
    "    \n",
    "    '''\n",
    "    \n",
    "    connection = sqlite3.connect('./data/SNP_human.db')\n",
    "\n",
    "    connection.execute('''CREATE TABLE IF NOT EXISTS SNP_data ( \n",
    "                           SNP_Name TEXT PRIMARY KEY, \n",
    "                           Alleles TEXT, \n",
    "                           Position TEXT, \n",
    "                           Chr TEXT,\n",
    "                           dbSNP_Build TEXT,\n",
    "                           Assembly_Build TEXT)''')\n",
    "\n",
    "    connection.execute('''CREATE TABLE IF NOT EXISTS clin_SNP (\n",
    "                           SNP_Name TEXT,\n",
    "                           Clinical_Significance TEXT,\n",
    "                           N_Publications TEXT,\n",
    "                           FOREIGN KEY (SNP_Name) REFERENCES SNP_data (SNP_Name))''')\n",
    "\n",
    "    connection.commit()\n",
    "    connection.close()\n",
    "    #Done in file\n",
    "    genes_list = get_genes()\n",
    "\n",
    "    id_list = get_id_list(genes_list)\n",
    "    red_id_list = list(set([j for sub in id_list for j in sub]))\n",
    "\n",
    "    with open('./data/SNP_ids.txt', 'w') as output_file:\n",
    "        for rsid in red_id_list:\n",
    "            output_file.write(rsid + '\\n')\n",
    "    \n",
    "    return get_snp_info(red_id_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc137239",
   "metadata": {},
   "outputs": [],
   "source": [
    "create_SNP_db()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b1f2f33e",
   "metadata": {},
   "source": [
    "# SNP DB select"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9701e16",
   "metadata": {},
   "outputs": [],
   "source": [
    "connection = sqlite3.connect('./data/SNP_human.db')\n",
    "\n",
    "query = '''SELECT SNP_data.SNP_Name, SNP_data.Chr, SNP_data.Position, SNP_data.Alleles\n",
    "           FROM SNP_data\n",
    "           LEFT JOIN clin_SNP ON SNP_data.SNP_Name = clin_SNP.SNP_Name\n",
    "           WHERE clin_SNP.Clinical_Significance = 'Reported in ClinVar' AND clin_SNP.N_Publications != '0'\n",
    "'''\n",
    "\n",
    "res = connection.execute(query)\n",
    "\n",
    "result = res.fetchall()\n",
    "connection.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "632707d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e65a8069",
   "metadata": {},
   "outputs": [],
   "source": [
    "connection = sqlite3.connect('./data/SNP_human.db')\n",
    "\n",
    "query = '''SELECT Chr, COUNT(SNP_Name)\n",
    "           FROM SNP_data\n",
    "           GROUP BY Chr\n",
    "           HAVING COUNT(SNP_Name) BETWEEN 2 AND 20\n",
    "'''\n",
    "\n",
    "res = connection.execute(query)\n",
    "\n",
    "result = res.fetchall()\n",
    "connection.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e98a1fe5",
   "metadata": {},
   "outputs": [],
   "source": [
    "result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "927f2122",
   "metadata": {},
   "outputs": [],
   "source": [
    "connection = sqlite3.connect('./data/SNP_human.db')\n",
    "\n",
    "query = '''BEGIN TRANSACTION;\n",
    "\n",
    "           DELETE FROM SNP_data\n",
    "           WHERE SNP_data.SNP_Name in (\n",
    "           SELECT SNP_Name FROM clin_SNP WHERE clin_SNP.Clinical_Significance = 'Not Reported in ClinVar');\n",
    "           \n",
    "           DELETE FROM clin_SNP\n",
    "           WHERE clin_SNP.Clinical_Significance = 'Not Reported in ClinVar'\n",
    "'''\n",
    "connection.execute(query)\n",
    "connection.commit()\n",
    "connection.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
